toolkit/tools/pkg/licensecheck/licensecheck.go

```diff
@@ -0,0 +1,452 @@
+	if err != nil {
+		newLicenseChecker.CleanUp()
+		err = fmt.Errorf("failed to initialize chroot. Error:\n%w", err)
+		return nil, err
+	}
+
+	if exceptionFilePath != "" {
+		newLicenseChecker.exceptions, err = LoadLicenseExceptions(exceptionFilePath)
+		if err != nil {
+			newLicenseChecker.CleanUp()
+			err = fmt.Errorf("failed to load license exceptions. Error:\n%w", err)
+			return nil, err
+		}
+	}
+
+	return newLicenseChecker, err
+}
+
+// CleanUp tears down the chroot. If the chroot was created it will be cleaned up. Reset the struct to its initial state.
+func (l *LicenseChecker) CleanUp() error {
+	if l.simpleToolChroot != nil {
+		err := l.simpleToolChroot.CleanUp()
+		if err != nil {
+			return fmt.Errorf("failed to cleanup chroot. Error:\n%w", err)
+		}
+		l.simpleToolChroot = nil
+	}
+	return nil
+}
+
+// CheckLicenses will scan all .rpm files in the chroot for bad licenses. The results can be accessed with FormatResults() or GetAllResults().
+func (l *LicenseChecker) CheckLicenses() error {
+	if l.simpleToolChroot == nil {
+		return fmt.Errorf("chroot has not been initialized")
+	}
+
+	l.results = []LicenseCheckResult{}
+
+	err := l.simpleToolChroot.RunInChroot(func() (searchErr error) {
+		l.results, searchErr = l.runLicenseCheckInChroot()
+		return searchErr
+	})
+	if err != nil {
+		return fmt.Errorf("failed to search for licenses. Error:\n%w", err)
+	}
+
+	for _, result := range l.results {
+		if result.err != nil {
+			logger.Log.Errorf("Failed to search %s. Error: %v", result.RpmPath, result.err)
+		}
+	}
+
+	// Sort the results by RPM path
+	// This is done to ensure that the output is deterministic
+	sort.Slice(l.results, func(i, j int) bool {
+		return l.results[i].RpmPath < l.results[j].RpmPath
+	})
+	return nil
+}
+
+// GetAllResults returns the results of the search, split into:
+// - All results: Every scan result
+// - Any result that has at least one warning
+// - Any result that has at least one error
+// There may be overlap between the lists. If pedantic is true, warnings will be treated as errors.
+func (l *LicenseChecker) GetAllResults(pedantic bool) (all, warnings, errors []LicenseCheckResult) {
+	all = []LicenseCheckResult{}
+	warnings = []LicenseCheckResult{}
+	errors = []LicenseCheckResult{}
+	for _, result := range l.results {
+		all = append(all, result)
+		if result.HasBadResult() || (pedantic && result.HasWarningResult()) {
+			errors = append(errors, result)
+		}
+		if result.HasWarningResult() && !pedantic {
+			warnings = append(warnings, result)
+		}
+	}
+	return all, warnings, errors
+}
+
+// FormatResults formats the results of the search to a string. Results will be ordered as follows:
+// - Packages with warnings only, sorted alphabetically
+// - Packages with errors (and possibly warnings), sorted alphabetically
+// If pedantic is true, warnings will be treated as errors.
+func (l *LicenseChecker) FormatResults(pedantic bool) string {
```

Comment:
I'm going to be working next on integrating this into the build scheduler / image build. I had hoped to re-use it there. If it doesn't work out I'll move it to the cmdline package.

Reply:
+1 to @cwize1's comment. It feels like we're mixing in additional responsibilities into the `LicenseChecker`. In addition to checking the licenses, it also deals with formatting the results, making it more tightly coupled to a single use case. I think it would make the code more flexible/modular, if we moved the formatting logic to a separate element, which would for instance accept the set of results as input. This way it's also going to be easier to extend the "formatters", if we need different ones in the future.